select

mikusp opened the pull request:
Select for sockets (and regular files on Linux)

mikusp committed:
select

StevenMiller123 commented on src/core/libraries/kernel/file_system.cpp:
```diff
@@ -1068,6 +1076,119 @@ s32 PS4_SYSV_ABI sceKernelUnlink(const char* path) {
+#endif
+        LOG_ERROR(Kernel_Fs, "native select call failed with {} ({})", error,
+                  Common::NativeErrorToString(error));
+        SetPosixErrno(error);
```
This still won't work properly on Windows, since winsock uses their net error codes here, and SetPosixErrno was designed around platform-specific errno values instead.

georgemoralis committed:
select for windows

StevenMiller123 commented on src/core/libraries/kernel/file_system.cpp:
```diff
@@ -1068,6 +1076,236 @@ s32 PS4_SYSV_ABI sceKernelUnlink(const char* path) {
+
+    HANDLE h = reinterpret_cast<HANDLE>(_get_osfhandle(fd)); // Convert to HANDLE
+    return h;
+}
```
This needs to be encased in an ifdef, since HANDLE is a Windows type.

georgemoralis committed:
fixed windows only function

georgemoralis committed:
windows error converts

georgemoralis committed:
fixed up

georgemoralis committed:
fixed compile

georgemoralis committed:
another implementation for windows

georgemoralis committed:
Merge branch 'main' into select

georgemoralis committed:
draft rewrite for windows

georgemoralis committed:
implementation for windows

georgemoralis committed:
added some debugging info

georgemoralis committed:
more debugging

georgemoralis committed:
extensive log

StevenMiller123 commented on src/core/libraries/kernel/file_system.cpp:
```diff
@@ -1067,6 +1075,324 @@ s32 PS4_SYSV_ABI sceKernelUnlink(const char* path) {
+    if (writefds)
+        FD_ZERO_POSIX(writefds);
+    if (exceptfds)
+        FD_ZERO_POSIX(exceptfds);
```
In games that only supply device fds, zeroing out the fd_set_posix parameters here will result in the function returning with empty output fds despite returning 0. Probably could be solved by adding the device fds to the `*_ready` sets above, then adding them back in alongside the regular file fds.
Addressing this problem will fix a regression in Cyberpunk 2077.

StevenMiller123 reviewed (changes requested):
I only ran into the one regression noted here with Cyberpunk 2077, everything else seems to behave fine.

mikusp committed:
Merge pull request #1 from StevenMiller123/select-pr

Windows: Add device files to output fd_sets

Commit:
Windows: Add device files to output fd_sets

Cyberpunk 2077 breaks without this.
Also added some formatting changes, to bring the implementation closer to our typical coding standards.

Commit:
Clang

Commit:
Formatting + cleanup

Removed some excessive logging used for debugging, and did some cleanup on the non-Windows implementation.
